More documentation changes in metrics (#820)
IBM/unitxt: docs/docs/adding_metric.rst

Metric Outputs
---------------

> By default, each metric provides scores for each instance separately and global scores over all instances together.

Reviewer: ```suggestion
By default, each metric provides scores for each instance separately and global aggregated scores over all instances together.
```

The output of the metrics is a nested dictionary per instance.

The scores calculated on instance `i` by itself are found in `results[i]["score"]["instance"]`. 
The global scores calculated over all instances are found in `results[i]["score"]["global"]`. 
Note the global scores are the same in all instances, so usually `results[0]["score"]["global"]` is used to get the global scores.

A metric could return multiple scores, but it should always return a field called `score` with the main score of the metric,

> and `score_name` which the name of the score.

Reviewer: ```suggestion
and `score_name` which the name of the main score.
```

Metric Base Classes
-------------------

As described in the previous section, a metric generate a set of scores per instance (called `instance` scores),
and a set of scores over all instances (called `global` scores).

Unitxt has several base classes :ref:`Metric <metrics>` class that simplify the creation of metrics, depending on how the
scores are calculated.

> InstanceMetric - Class for metrics in which the global scores are be calculated by aggregating the instance scores.
> Typically, the global score is the average of all instance scores.  InstanceMetric first evaluates each instance separately,
> and then aggregate the instances score.   Some examples of instance metrics are Accuracy, TokenOverlap, CharEditDistance.  

Reviewer: ```suggestion
`InstanceMetric` - Class for metrics in which the global scores are be calculated by aggregating the instance scores.
Typically, the global score is the average of all instance scores.  `InstanceMetric` first evaluates each instance separately,
and then aggregate the instances score.  Some examples of instance metrics are `Accuracy`, `TokenOverlap`, `CharEditDistance`.  
```

> BulkInstanceMetric - Similar to InstanceMetric , it is for metrics in which the globals score can be calculated by aggregating the instance scores.  However,
> due to implementation efficiently reasons, it's better to run them in bulk (for example, when using LLMs during score calculations).
> BulkInstanceMetric runs on a batch of instances each time, but then aggregate the instance scores as before.
> Some examples of bulk instance metrics are SentenceBert, Reward.

Reviewer: ```suggestion
`BulkInstanceMetric` - Similar to I`nstanceMetric` , it is for metrics in which the globals score can be calculated by aggregating the instance scores.  However,
due to implementation efficiently reasons, it's better to run them in bulk (for example, when using LLMs during score calculations).
`BulkInstanceMetric` runs on a batch of instances each time, but then aggregate the instance scores as before.
Some examples of bulk instance metrics are `SentenceBert`, `Reward`.
```

> GlobalMetric - Class for metrics for which the global scores must be calculated over all the instances together.
> Some examples of global metrics are f1, Spearman, Kendall Tau.  Note that by default global metrics are executed once per instance 

Reviewer: ```suggestion
`GlobalMetric` - Class for metrics for which the global scores must be calculated over all the instances together.
Some examples of global metrics are `f1`, `Spearman`, `Kendall Tau`.  Note that by default global metrics are executed once per instance 
```

to generate per instance scores, and then once again over all instances together.   So if there are 100 instances, 
it will first be called 100 times , each on a single instance, and then one time on all 100 instances.  

> Instance scores of GlobalMetrics are useful for error-analysis. Consider f1 score, for example. 

Reviewer: ```suggestion
Instance scores of `GlobalMetrics` are useful for error-analysis. Consider f1 score, for example. 
```

            result = { self.main_score :  spearman_coeff }
            return result

> 1. Calculating confidence internals for global metricscan be costly if each invocation of the metric takes a long time.

Reviewer: ```suggestion
1. Calculating confidence intervals for global metrics can be costly if each invocation of the metric takes a long time.
```

        )
        add_to_catalog(metric, "metrics.token_overlap_with_context", overwrite=True)

Adding a Hugginface metric
----------------------------

Unitxt provides a simple way to wrap existing Huggingface without the need to write code.
This is done using the predefined HuggingfaceMetric class.

.. code-block:: python

    metric = HuggingfaceMetric(
        hf_metric_name="bleu",  # The name of the metric in huggingface
        main_score="bleu",      # The main score (assumes the metric returns this score name)
        prediction_type="str"   # The type of the prediction and references (note that by default references are a list of the prediction_type)
    )
    add_to_catalog(metric, "metrics.bleu", overwrite=True)

By default, the HuggingfaceMetric wrapper passes the only the `predictions` and `references` fields to 

> the metrics.  You can also pass fields from the task_data inputs, by specifying `hf_additional_input_fields`.

Reviewer: ```suggestion
the metrics. You can also pass fields from the task_data inputs, by specifying `hf_additional_input_fields`.
```

For example:

 .. code-block:: python

    metric = HuggingfaceMetric(
        ...
        hf_additional_input_fields_pass = ["num1","num2"], # passes the task's num1 and num2 fields 
        ...
        

>      )    

Reviewer: ```suggestion
    )    
```

In the above example, the `num1` and `num2`fields are passed as lists of values to the metric

> (each element in the list corresponds to an instance).  If you want to pass a scalar (single) value to the metric

Reviewer: ```suggestion
(each element in the list corresponds to an instance). If you want to pass a scalar (single) value to the metric
```

you can use:

.. code-block:: python

    metric = HuggingfaceMetric(
        ...
        hf_additional_input_fields_pass_one_value=["tokenize"],
        ...
     )   
   

This assumes the field has the same value is in all instances.

> Note that Huggingface metrics are independent the tasks they are used for, and receive arbitrary types of predictions, references, and additional

Reviewer: ```suggestion
Note that Huggingface metrics are independent from the tasks they are used for, and receive arbitrary types of predictions, references, and additional
```
Is that what you meant? this section is a bit unclear
